macOS/LocalPackages/NetworkProtectionMac/Sources/NetworkProtectionUI/Views/TunnelControllerView/TunnelControllerViewModel.swift

```diff
@@ -264,17 +265,28 @@ public final class TunnelControllerViewModel: ObservableObject {
             }
 
             return self.internalIsRunning
-        } set: { newValue in
+        } set: { [weak self] newValue in
+            guard let self else { return }
+
             guard newValue != self.internalIsRunning else {
                 return
             }
 
-            self.internalIsRunning = newValue
-
-            if newValue {
-                self.startNetworkProtection()
-            } else {
-                self.stopNetworkProtection()
+            Task {
+                // When turning OFF the VPN we let the parent app offer an alternative
+                // UI/UX to prevent it.  If they return `false` in `willStopVPN` the VPN
+                // will not be stopped.
+                if !newValue, await !self.uiActionHandler.willStopVPN() {
+                    return
+                }
+
+                self.internalIsRunning = newValue
```

Comment:
I needed to move this inside a task so it will wait until the above `willStopVPN()` call completed.  This moves the toggle to the left when the user clicks, so we want to avoid doing that until the alert has been shown and the user has picked an option.